astropy/convolution/tests/test_convolve_fft.py

```diff
@@ -342,20 +343,66 @@ def test_uniform_3_withnan(self, boundary, nan_treatment,
         if nan_treatment == 'interpolate':
             answer_key += '_interpnan'
 
-        posns = np.where(np.isfinite(z))
+        posns = np.isfinite(z)
 
-        assert_floatclose(z[posns], answer_dict[answer_key][posns])
+        answer = answer_dict[answer_key][posns]
 
-    def test_nan_fill(self):
+        # check that fill is set and that the 1'th position that was originally
+        # NaN is included in the check
+        if (nan_treatment == 'fill') and posns[1]:
+            # we fill the center with the sum of the input array divided by
+            # three, since we've now pre-filled the center value with zero
+            answer[1] = 4 / (3. if normalize_kernel else 1.)
+
+        assert_floatclose(z[posns], answer)
+
+    def test_nan_interpolate(self):
 
         # Test masked array
         array = np.array([1., np.nan, 3.], dtype='float64')
         kernel = np.array([1, 1, 1])
         masked_array = np.ma.masked_array(array, mask=[0, 1, 0])
+
         result = convolve_fft(masked_array, kernel, boundary='fill',
+                              nan_treatment='interpolate',
                               fill_value=np.nan)
+
         assert_floatclose(result, [1, 2, 3])
 
+    def test_nan_fill(self):
+        # regression for #8121
+
+        # Test masked array
+        array = np.array([1., np.nan, 3.], dtype='float64')
+        kernel = np.array([1, 1, 1])
+
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore', RuntimeWarning)
+            result = convolve_fft(array, kernel, boundary='fill',
+                                  nan_treatment='fill',
+                                  fill_value=0)
+
+        # note that, because fill_value also affects boundary='fill', the edge
+        # pixels are treated as zero rather than being ignored.
```

Comment:
Right, this is the bug, it shouldn't happen this way.

Reply:
Yeah, this isn't so much a bug as a design failure, as highlighted in #8127

Reply:
Probably, however, it's not always the case that the developer gets to define what is and isn't a bug. I.e. the artist of modern abstract art doesn't get to completely define its interpretation. 

If the behaviour and functionality don't intuitively match the semantics (or the documentation (loophole)) then it is a bug. Even if the fix is just readjusting the semantics by documenting it as a feature (hence the loophole), it's still a bug until fixed.

Reply:
I'm going to stick with "design failure" because neither the documentation nor the implementation actually make sense.  As you pointed out, `fill_value` is incorrectly tied to both `boundary` and `nan_treatement`, which results in some nonsense parameter combinations.

That's said, this is all semantics, we have a bunch of crap that needs to be cleaned up, and I'm just arguing for punting the fixes related to this particular line to #8127, since it more accurately and completely describes the problem.

There is an open question here, though - should this PR be merged before dealing with the bigger issues in 8127, or should we try to get everything right all at once?  I'm open to suggestions; I don't believe there are many/any users trying to use this corner of the convolution code yet, which gives us some flexibility, but we want it all to be right as soon as possible...

Reply:
I'm cool to defer this particular point to #8127. I think it might be better and safer to address this piece by piece. I'm already trying to untangle the normalization and nan treatment as I did for convolve, so we already have a few parallel jobs.

Reply:
:+1:  Thanks for doing this. 

Just to give you some background: these problems mostly come from `convolve` and `convolve_fft` being developed fully independently (convolve by Tom, convolve_fft by me), then trying to match their call specifications.  In trying to make `convolve_fft` match `convolve`'s arguments more closely, I clearly missed/ignored several use cases that I had never considered.  It's looking more likely that I caught all of the inverse, i.e., things convolve_fft did that needed to be ported over to convolve.